src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,197 @@
+import React, { useEffect, useRef, useState } from 'react';
+import classNames from 'classnames';
+
+import { deleteTodoOnServer, updateTodoOnServer } from '../../api/todos';
+
+import { ErrorType } from '../../enums/ErrorType';
+import { Todo } from '../../types/Todo';
+import { OnChangeFunc } from '../../types/OnChangeFunc';
+
+type Props = {
+  todo: Todo;
+  isLoading: boolean;
+  onDeleteTodo?: (todoId: number) => void;
+  onChangeTodo?: OnChangeFunc;
+  showError?: (errorType: ErrorType) => void;
+  hideError?: () => void;
+};
+
+export const TodoItem: React.FC<Props> = React.memo(
+  ({
+    todo,
+    isLoading,
+    onDeleteTodo = () => {},
+    onChangeTodo = () => {},
+    showError = () => {},
+    hideError = () => {},
+  }) => {
+    const { id, title, completed } = todo;
+
+    const [isWaiting, setIsWaiting] = useState(false);
+    const [isEditAllowed, setIsEditAllowed] = useState(false);
+    const [editedTitle, setEditedTitle] = useState(title);
+
+    const editFormRef = useRef<HTMLInputElement | null>(null);
+
+    useEffect(() => {
+      if (editFormRef.current && isEditAllowed) {
+        editFormRef.current.focus();
+      }
+    }, [isEditAllowed]);
+
+    const handleDeleteTodo = async (onError?: () => void) => {
+      hideError();
+      setIsWaiting(true);
+
+      try {
+        await deleteTodoOnServer(id);
+
+        onDeleteTodo(id);
+      } catch {
+        showError(ErrorType.Delete);
+        setIsWaiting(false);
+
+        if (onError) {
+          onError();
+        }
+      }
+    };
+
+    const handleChangeTodo: OnChangeFunc = async (
+      todoId,
+      propName,
+      newPropValue,
+      onError,
+    ) => {
+      hideError();
+      setIsWaiting(true);
+
+      try {
+        await updateTodoOnServer(todoId, { [propName]: newPropValue });
+
+        onChangeTodo(todoId, propName, newPropValue);
+      } catch {
+        showError(ErrorType.Update);
+
+        if (onError) {
+          onError();
+        }
```

Comment:
this can be simplified 
```suggestion
        onError?.();
```